### PR TITLE
[Kineto plugin] Implement kineto plugin and register

### PR DESCRIPTION
**Pull request by zejun-chen**
[Kineto plugin] Implement kineto plugin and register xpu to torch kineto
RFC: https://github.com/intel-innersource/frameworks.ai.pytorch.private-gpu/issues/234 (Need to revise)

TODO:
- [ ] improve the RFC draft
- [ ] check building on public torch 2.3
- [ ] make lint happy
- [ ] cherry pick to the private torch 2.3
- [ ] verify the correctness with IPEX

### CODE REVIEW

File: c10/kineto_plugin/register_api.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  #ifdef USE_##TYPE                                           \
+  libkineto::api().registerProfilerFactory(                   \
+      []() -> std::unique_ptr<IActivityProfiler> {            \
+        return std::make_unique<TYPE##ActivityProfiler>(); \
+      });                                                     \
+  #endif
+
+// register new kineto backends in kineto plugin
+void registerKinetoPluginProfiler() {
+  REGISTER_KINETO_PLUGIN(XPU);
```
Comment (gujinghui) [inline c10/kineto_plugin/register_api.cpp:20]: It's not a good design. If we are going to provide a common interface to register a kineto backend, it should not depend on device specific code.
Can we have a more general design here? Reference to allocator design?

**Comment by zejun-chen** [comment]
1. Check the init callback timing. Make sure register before callback. When the prepare_trace is called?
2. No more new library. C10 for register method and torch_xpu for xpu register
3. CMake make dependency for needed header file

File: c10/xpu/XPUFunctions.cpp [hunk 2]
```diff
@@ -154,6 +160,16 @@ DeviceIndex get_device_idx_from_pointer(void* ptr) {
       std::distance(gDevicePool.devices.begin(), it));
 }
 
+DeviceIndex get_device_idx_from_uuid(const uint8_t device_uuid[16]) {
+  std::array<unsigned char, 16> key;
+  memcpy(key.data(), device_uuid, 16);
+  auto it = std::find(gDevicePool.uuids.begin(), gDevicePool.uuids.end(), key);
+  if (it != gDevicePool.uuids.end()) {
+    return static_cast<int64_t>(std::distance(gDevicePool.uuids.begin(), it));
```
Comment (guangyey) [inline c10/xpu/XPUFunctions.cpp:168]:
```suggestion
    return static_cast<DeviceIndex>(std::distance(gDevicePool.uuids.begin(), it));
```

File: c10/xpu/XPUFunctions.cpp [hunk 3]
```diff
@@ -154,6 +160,16 @@ DeviceIndex get_device_idx_from_pointer(void* ptr) {
...
+  }
+  return static_cast<int64_t>(-1);
```
Comment (guangyey) [inline c10/xpu/XPUFunctions.cpp:170]:
```suggestion
  return static_cast<DeviceIndex>(-1);
```

File: c10/kineto_plugin/register_api.cpp (hunk 1 above)
Comment (guangyey) [reply c10/kineto_plugin/register_api.cpp:20]: I see. `registerKinetoPluginProfiler` will be called in `kineto_shim.cpp` file as RAII. Refer to allocator design(new it and leak memory intendedly) can solve timing issue.

**Pull request closed without merging**